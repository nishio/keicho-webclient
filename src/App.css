.App {
  text-align: left;
}

textarea {
  width: calc(100% - 10px);
  margin: 5px;
  box-sizing: border-box;
  font-size: 20px;
  border-radius: 17px;
  padding: 0 10px;
  word-break: break-all;
}
