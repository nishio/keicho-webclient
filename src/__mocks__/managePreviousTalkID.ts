// no previoud id
export const getPreviousTalkID = (): Promise<string> => {
  return Promise.resolve("");
};
export const MOCK_PREVIOUS_ID_EXISTS = (): Promise<string> => {
  return Promise.resolve("test");
};

export const updatePreviousTalkID = (currentTalkID: string): void => {};
