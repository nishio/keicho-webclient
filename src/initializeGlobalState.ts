import { setGlobal } from "reactn";
import { INITIAL_LOGS, PRESET_LOGS, USE_PRESET } from "./PRESET_MESSAGES";

const INITIAL_GLOBAL_STATE = {
  TalkID: "",
  logs: USE_PRESET ? PRESET_LOGS : INITIAL_LOGS,
  lastKeywords: [] as string[],
  otherKeywords: [] as string[],
  canInput: true,
  previousTalkID: "",
};

export const initializeGlobalState = () => {
  setGlobal(INITIAL_GLOBAL_STATE);
};

type TYPE_GLOBAL_STATE = typeof INITIAL_GLOBAL_STATE;

declare module "reactn/default" {
  export interface State extends TYPE_GLOBAL_STATE {}
}
