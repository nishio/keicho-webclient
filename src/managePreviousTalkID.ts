import { localDB } from "./localDB";

export const getPreviousTalkID = (): Promise<string> => {
  return localDB.talks
    .orderBy("id")
    .reverse()
    .limit(1)
    .toArray()
    .then((x) => {
      if (x[0] !== undefined) {
        return Promise.resolve(x[0].TalkID);
      } else {
        return Promise.resolve("");
      }
    });
};
export const updatePreviousTalkID = (currentTalkID: string): void => {
  localDB.talks.add({ TalkID: currentTalkID });
};
