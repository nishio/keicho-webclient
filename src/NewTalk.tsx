import React, { KeyboardEventHandler, useEffect, useState } from "react";
import Menu from "./Menu";
import { TextareaAutosize } from "@material-ui/core";
import { ChatLine } from "./ChatContents";
import { PRESET_LOGS } from "./PRESET_LOGS";
import { USE_PRESET, INITIAL_LOGS, APIROOT, TalkID } from "./App";

export const NewTalk = () => {
  const [logs, setLogs] = useState(USE_PRESET ? PRESET_LOGS : INITIAL_LOGS);

  useEffect(() => {
    fetch(APIROOT + "web/create", {
      mode: "cors",
      method: "GET",
    })
      .then((response) => {
        return response.text();
      })
      .then((text) => {
        TalkID = text;
      });
  }, []);

  const onKeyPress: KeyboardEventHandler<HTMLTextAreaElement> = (e) => {
    if (e.key === "Enter") {
      const target = e.target as HTMLTextAreaElement;
      const text = target.value;
      target.value = "";
      e.preventDefault();
      if (text !== "") {
        const newLogs = [...logs, { text: text, user: true }];
        setLogs(newLogs);
        // send to server
        const data = { user: "nobody", talk: TalkID, text: text };
        fetch(APIROOT + "web/", {
          mode: "cors",
          method: "POST",
          body: JSON.stringify(data),
          headers: {
            "Content-Type": "application/json",
          },
        }).then((response) => {
          console.log(response);
          response.json().then((data) => {
            setLogs([...newLogs, { text: data.text, user: false }]);
          });
        });
      }
    }
  };

  return (
    <div className="App">
      <Menu />
      <ChatLine logs={logs}></ChatLine>

      <TextareaAutosize
        aria-label="empty textarea"
        placeholder=""
        onKeyPress={onKeyPress}
      />
    </div>
  );
};
