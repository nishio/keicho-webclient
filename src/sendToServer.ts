import {
  BOT_IS_SLEEPING,
  ERROR_ON_SERVER,
  BOT_IS_THINKING,
} from "./PRESET_MESSAGES";
import { getGlobal, setGlobal } from "reactn";
import { APIROOT } from "./App";
import * as Sentry from "@sentry/browser";
import { updateTalk } from "./updateTalk";

type TLogs = { text: string; user: boolean }[];

export function sendToServer(text: string, newLogs: TLogs) {
  const g = getGlobal();
  if (g.TalkID !== "") {
    const data = { user: "nobody", talk: g.TalkID, text: text };
    updateTalk(g.TalkID, text);
    setGlobal({ logs: [...newLogs, BOT_IS_THINKING] });

    const transaction = Sentry.startTransaction({ name: "getNewTalkID" });
    const span = transaction.startChild({ op: "getNewTalkID" });

    return fetch(APIROOT + "web/", {
      mode: "cors",
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((response) => {
        return response.json();
      })
      .then((data) => {
        if (data.text !== "") {
          return _gotResponse(newLogs, data);
        }
      })
      .then(() => {
        span.finish();
        transaction.finish();
      })
      .catch(() => {
        Sentry.captureMessage("ERROR_ON_SERVER");
        setGlobal({
          logs: [...newLogs, ERROR_ON_SERVER],
          canInput: false,
        });
      });
  } else {
    // bot is sleeping
    setGlobal({ logs: [...newLogs, BOT_IS_SLEEPING] });

    return new Promise((resolve) => setTimeout(resolve, 1000)).then(() => {
      sendToServer(text, newLogs);
    });
  }
}

export const _gotResponse = (newLogs: TLogs, data: any) => {
  return setGlobal({
    logs: [...newLogs, { text: data.text, user: false }],
    buttons: data.buttons,
    canInput: data.can_input,
  });
};
