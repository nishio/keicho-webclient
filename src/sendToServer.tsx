import {
  BOT_IS_SLEEPING,
  ERROR_ON_SERVER,
  BOT_IS_THINKING,
} from "./PRESET_MESSAGES";
import { getGlobal, setGlobal } from "reactn";
import { APIROOT } from "./App";

export function sendToServer(
  text: string,
  newLogs: { text: string; user: boolean }[]
) {
  const g = getGlobal();
  if (g.TalkID !== "") {
    const data = { user: "nobody", talk: g.TalkID, text: text };
    setGlobal({ logs: [...newLogs, BOT_IS_THINKING] });

    fetch(APIROOT + "web/", {
      mode: "cors",
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((response) => {
        response.json().then((data) => {
          if (data.text !== "") {
            setGlobal({
              logs: [...newLogs, { text: data.text, user: false }],
              lastKeywords: data.last_kw,
              otherKeywords: data.other_kw,
            });
          }
        });
      })
      .catch(() => {
        setGlobal({
          logs: [...newLogs, ERROR_ON_SERVER],
          canInput: false,
        });
      });
  } else {
    // bot is sleeping
    setGlobal({ logs: [...newLogs, BOT_IS_SLEEPING] });

    setTimeout(() => {
      sendToServer(text, newLogs);
    }, 1000);
  }
}
