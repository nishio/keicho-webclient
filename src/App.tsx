import React, { useEffect } from "react";
import "./App.css";
import db from "./FirestoreIO";
import { NewTalk } from "./NewTalk";

const USE_LOCAL_SERVER = false;
const IS_PRODUCTION = process.env.NODE_ENV === "production";
export const APIROOT =
  IS_PRODUCTION || !USE_LOCAL_SERVER
    ? "https://keicho.herokuapp.com/api/"
    : "http://localhost:5000/api/";

export const USE_PRESET = false;
export const INITIAL_LOGS = [
  { text: "あなたはこの会話で何が起きて欲しいですか？", user: false },
];

const ShowLog = () => {
  useEffect(() => {
    const key = "I9IlVW6Xt58NlNHAjdfq";
    console.log("reading", key);
    db.collection("yagokoro_env")
      .doc(key)
      .get()
      .then((doc: any) => {
        if (doc.exists) {
          const data = doc.data();
          console.log(data);
          const x = JSON.parse(data.json);
          console.log(x);
        } else {
          // doc.data() will be undefined in this case
          console.log("No such document!");
        }
      });
  }, []);
  return <br />;
};

function App() {
  const hash = new URLSearchParams(window.location.hash.substring(1));
  if (hash.has("talk")) {
    return <ShowLog />;
  } else {
    return <NewTalk />;
  }
}

export default App;
