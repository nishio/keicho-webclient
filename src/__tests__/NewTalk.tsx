import { unmountComponentAtNode } from "react-dom";
import { initializeGlobalState } from "../initializeGlobalState";
import { act, fireEvent, render, screen } from "@testing-library/react";
import App from "../App";
import * as managePreviousTalkIDModule from "../managePreviousTalkID";
import { getNewTalkIDPromise, sendToServerPromise } from "../NewTalk";
import { mockUseState } from "./mockUseState";
import * as sendToServerModule from "../sendToServer";
import userEvent from "@testing-library/user-event";
jest.mock("../managePreviousTalkID");
jest.mock("../getNewTalkIDFromServer");

jest.spyOn(window, "scrollTo").mockImplementation(() => {});

let container: HTMLDivElement;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
  act(() => {
    initializeGlobalState();
  });
  mockUseState();
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
});

test("render", async () => {
  // render when previousTalkID does not exist
  jest
    .spyOn(managePreviousTalkIDModule, "getPreviousTalkID")
    .mockResolvedValue("");

  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
  fireEvent.click(screen.getByLabelText("menu"));
  await getNewTalkIDPromise;
  expect(screen.queryByText("Re-enter to Last Talk")).toBeNull();

  // render when previousTalkID exists
  jest
    .spyOn(managePreviousTalkIDModule, "getPreviousTalkID")
    .mockResolvedValue("test");
  render(<App />);
  await getNewTalkIDPromise;
  expect(screen.queryByText("Re-enter to Last Talk")).not.toBeNull();
});

test("enter", async () => {
  jest
    .spyOn(managePreviousTalkIDModule, "getPreviousTalkID")
    .mockResolvedValue("");
  jest
    .spyOn(sendToServerModule, "sendToServer")
    .mockImplementation(async () => {
      sendToServerModule._gotResponse([{ user: true, text: "aaa" }], {
        text: "bbb",
        last_kw: ["a"],
        other_kw: ["b"],
      });
    });

  const { container } = render(<App />);
  const textarea = container.querySelector("[id=textarea]");
  expect(textarea!).not.toBeNull();

  act(() => {
    userEvent.type(textarea!, "Hello, World!{enter}");
  });
  expect(textarea!).toHaveValue("");
  await sendToServerPromise;
  expect(container).toMatchSnapshot();
});
