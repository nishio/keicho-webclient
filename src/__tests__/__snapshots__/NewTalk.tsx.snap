// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enter 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="makeStyles-root-3"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h6
            class="MuiTypography-root makeStyles-title-4 MuiTypography-h6"
          >
            Keicho
          </h6>
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd"
            id="menu"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
    <div
      class="line__container"
    >
      <div
        class="line__contents"
      >
        <div
          class="line__right"
        >
          <div
            class="text"
          >
            aaa
          </div>
        </div>
        <div
          class="line__left"
        >
          <figure>
            <img
              alt="nisbot"
              src="robot.png"
            />
          </figure>
          <div
            class="line__left-text"
          >
            <div
              class="text"
            >
              bbb
            </div>
          </div>
        </div>
      </div>
    </div>
    <textarea
      aria-label="textarea"
      id="textarea"
      placeholder=""
      rows="1"
      style="height: -4px; overflow: hidden;"
    />
    <textarea
      aria-hidden="true"
      readonly=""
      style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0);"
      tabindex="-1"
    />
    <button
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        üôÅ
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        üôÅ
        a
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        üëç
        b
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <hr
      id="bottom"
    />
  </div>
</div>
`;

exports[`render 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="makeStyles-root-1"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h6
            class="MuiTypography-root makeStyles-title-2 MuiTypography-h6"
          >
            Keicho
          </h6>
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd"
            id="menu"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
    <div
      class="line__container"
    >
      <div
        class="line__contents"
      >
        <div
          class="line__left"
        >
          <figure>
            <img
              alt="nisbot"
              src="robot.png"
            />
          </figure>
          <div
            class="line__left-text"
          >
            <div
              class="text"
            >
              „ÅÇ„Å™„Åü„ÅØ„Åì„ÅÆ‰ºöË©±„Åß‰Ωï„ÅåËµ∑„Åç„Å¶Ê¨≤„Åó„ÅÑ„Åß„Åô„ÅãÔºü
            </div>
          </div>
        </div>
      </div>
    </div>
    <textarea
      aria-label="textarea"
      id="textarea"
      placeholder=""
      rows="1"
      style="height: -4px; overflow: hidden;"
    />
    <textarea
      aria-hidden="true"
      readonly=""
      style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0);"
      tabindex="-1"
    />
    <button
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        üôÅ
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <hr
      id="bottom"
    />
  </div>
</div>
`;
